plugins/cpu/fu-plugin-cpu.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2020 Mario Limonciello <mario.limonciello@dell.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-plugin-vfuncs.h"
+#include "fu-hash.h"
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+	gsize length;
+	g_autofree gchar *data = NULL;
+	g_auto(GStrv) lines = NULL;
+	g_autofree gchar *processor = NULL;
+	g_autofree gchar *model_name = NULL;
+	g_autofree gchar *microcode = NULL;
+	g_autofree gchar *vendor = NULL;
+	guint64 physical_id = G_MAXUINT64;
+
+	if (!g_file_get_contents ("/proc/cpuinfo", &data, &length, error))
+		return FALSE;
+
+	lines = g_strsplit (data, "\n", 0);
```

Comment:
the purist in me would say split on `\n\n` and then pass each chunk into a fu_cpu_device_new(const gchar *section) where it gets parsed there. Then when you get the FuCpuDevice back you look at the physical ID and if it's the same as the `physical_id` then it gets unrefd and ignored. I don't think it's any more fragmentation compared to what you have here.